src/rpm/signature/traits.rs

```diff
@@ -8,63 +8,58 @@ use crate::errors::*;
 use std::fmt::Debug;
 use std::io::Read;
 
-pub mod algorithm {
-
-    pub trait Algorithm: super::Debug {}
-    /// currently only RSA is required
-    ///
-    /// Farsight for future algorithm extensions of rpm
-    /// without breaking the API
-    #[derive(Debug, Clone, Copy)]
-    #[allow(non_camel_case_types)]
-
-    pub struct RSA;
-
-    impl Algorithm for RSA {}
+#[derive(Clone, Copy, Debug)]
+pub enum AlgorithmType {
+    RSA,
+    EdDSA,
 }
 
 /// Signing trait to be implement for RPM signing.
-pub trait Signing<A>: Debug
+pub trait Signing: Debug
 where
-    A: algorithm::Algorithm,
     Self::Signature: AsRef<[u8]>,
 {
     type Signature;
     fn sign<R: Read>(&self, data: R) -> Result<Self::Signature, RPMError>;
+    fn algorithm(&self) -> AlgorithmType;
```

Comment:
So here's my thinking:

If we are just handing the Signer / Verifier raw bytes so that it can do the PGP key parsing, then it makes no sense to use a compile-time generic for the algorithm type.  Because when it parses the key, it can just return an error if the key is of the wrong type.  And since all the algorithms are handled by the same PGP implementation anyway, it would be silly to duplicate the code.

I'm approximately 100% sure we could make this more ergonomic still, but this is as far as I wanted to go in this PR, we can continue cleaning it up later.

Reply:
Do we really want automatic derivation of the algorithm from the key? That's my key (no pun intended) question here. I'd imagine that a user wants to use a specific key. The compile time guarantees could still be combined.

```
fn sign_with_key_given(self, key: ..) {
let key = key.parse()?;
match key.algorithm() {
Algorithm::RSA => signer.sign::<Algorithm::RSA>(),
// ..
}
```

now this guarantees compiletime consistency and only picks the path according to the convenience route involving algorithm derivation. People that do have the knowledge of the key type statically could maintain correctness by definition.

That's my thinking. Pushing out the runtime checks as far as possible.